### PR TITLE
Ansible OS cluster testing

### PR DESCRIPTION
**Pull request by peterchoo**
## Problem

Basic cluster testing must be performed manually, by following the Microsoft documentation for [SLES](https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/sap/sap-hana-high-availability#test-the-cluster-setup) and [RHEL](https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/sap/sap-hana-high-availability-rhel#test-the-cluster-setup).

## Description

This PR aims to provide an automated method for triggering and attempting to recover from failover tests for:

1. Manually migrating the Master node
1. Testing the Fence Agent reacts appropriately to the loss of the DB NIC
1. Failure of the Clustering Service

## TODO

- [x] Document helper script usage

## Pre-Requisites

None

## Commitment to Test

- Testing takes ~1 hour, mostly waiting for tests
- Reviewing all files

### Scenario 1: Code review

1. Review the [shell script](https://github.com/Centiq/sap-hana/pull/41/files?file-filters%5B%5D=.sh&file-filters%5B%5D=.yml#diff-8872431dd3c920d66445b6dcec6b5fa0R1)
   - [x] Follows util script conventions
   - [x] Self documenting
1. Review the [playbook](https://github.com/Centiq/sap-hana/pull/41/files?file-filters%5B%5D=.sh&file-filters%5B%5D=.yml#diff-9ce87543846298e19b389bead01b2411R1)
   - [x]  Code OK
1. Review the [Variables](https://github.com/Centiq/sap-hana/pull/41/files?file-filters%5B%5D=.sh&file-filters%5B%5D=.yml#diff-0280ff70931d47ae58e1c2e647481c3cR1)
   - [x] Variables are clearly defined
1. Review the task files
   - [x] Tasks are clearly named
   - [x] Tasks separation is clear and easily understood
1. Review the [documentation changes](https://github.com/Centiq/sap-hana/pull/41/files#diff-7b2ac2a41f894bd0e6577ab6721d0d9dR304)
   - [x] Instructions are clear

### Scenario 2: SLES testing

For this test, you must provide your public key to gain access to the RTI.

1. Connect to the RTI:\
   `ssh <redacted>`
1. Trigger the migration test:
   `~/sap-hana/deploy/v2/test_cluster.sh migrate`
1. Follow the onscreen instructions, migration takes approx 3-5 minutes
   - [x] Failover test passes
   - [x] Cluster is re-established
1. Trigger the Fencing Agent test:
   `~/sap-hana/deploy/v2/test_cluster.sh fence_agent`
1. Follow the onscreen instructions, migration takes approximately 15 minutes
   - [x] Master node is rebooted
   - [x] Failover test passes
   - [x] Cluster is re-established
1. Trigger the Service test:
   `~/sap-hana/deploy/v2/test_cluster.sh service`
1. Follow the onscreen instructions, failover takes approximately 15 minutes
   - [x] Master node is rebooted
   - [x] Failover test passes
   - [x] Cluster is re-established

### Scenario 3: RHEL testing

For this test, you must provide your public key to gain access to the RTI.

1. Connect to the RTI:\
   `ssh <redacted>`
1. Trigger the migration test:
   `~/sap-hana/deploy/v2/test_cluster.sh migrate`
1. Follow the onscreen instructions, migration takes approx 3-5 minutes
   - [x] Failover test passes
   - [x] Cluster is re-established
1. Trigger the Fencing Agent test:
   `~/sap-hana/deploy/v2/test_cluster.sh fence_agent`
1. Follow the onscreen instructions, migration takes approximately 15 minutes
   - [x] Master node is rebooted
   - [x] Failover test passes
   - [x] Cluster is re-established
1. Trigger the Service test:
   `~/sap-hana/deploy/v2/test_cluster.sh service`
1. Follow the onscreen instructions, failover takes approximately 15 minutes
   - [x] Master node is rebooted
   - [x] Failover test passes
   - [x] Cluster is re-established

## References

- [SLES Cluster Testing](https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/sap/sap-hana-high-availability#test-the-cluster-setup)
- [RHEL Cluster Testing](https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/sap/sap-hana-high-availability-rhel#test-the-cluster-setup)

### CODE REVIEW

File: deploy/v2/ansible/roles/test-failover/vars/main.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+cluster_detail_cmds:
+  Suse:
+    master_hostname: "SAPHanaSR-showAttr | grep PROMOTED | cut -d' ' -f1"
+    master_site_name: "SAPHanaSR-showAttr | grep PROMOTED | tr -s ' ' | cut -d' ' -f9"
+    secondary_hostname: "SAPHanaSR-showAttr | grep DEMOTED | cut -d' ' -f1"
+    msl_name: "crm status | grep '^ Master' | cut -d' ' -f4"
+    sid: "crm configure show | grep -m1 SID= | grep -oP 'SID=\\K([a-zA-Z0-9]{3})'"
+    instance_number: "crm configure show | grep -m1 InstanceNumber= | grep -oP 'InstanceNumber=\\K(\\d{2})'"
+    db_ip: "ip addr show eth1 | grep -m1 inet | grep -oP 'inet \K([0-9]+\.){3}[0-9]+'"
```
Comment (robfallows) [inline deploy/v2/ansible/roles/test-failover/vars/main.yml:29]:
```suggestion
    db_ip: "ip addr show eth1 | grep -m1 inet | grep -oP 'inet \\K([0-9]+\\.){3}[0-9]+'"
```

File: deploy/v2/ansible/roles/test-failover/vars/main.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  RedHat:
+    master_hostname: "pcs status --full | grep -B1 PROMOTED | grep -oP '^\\* Node \\K([a-zA-Z0-9\\-\\.]*)'"
+    master_site_name: "pcs status --full | grep -A10 PROMOTED | grep -m1 _site | grep -oP ': \\K([a-zA-Z0-9\\-\\.\\_]+)'"
+    secondary_hostname: "pcs status --full | grep -B1 DEMOTED | grep -oP '^\\* Node \\K([a-zA-Z0-9\\-\\.]*)'"
+    msl_name: "pcs resource show --full | grep -oP '^ Master: \\K.*'"
+    sid: "pcs resource show --full | grep -m1 SID | grep -oP 'SID=\\K([a-zA-Z0-9]{3})'"
+    instance_number: "pcs resource show --full | grep -m1 InstanceNumber | grep -oP 'InstanceNumber=\\K(\\d{2})'"
+    db_ip: "ip addr show eth1 | grep -m1 inet | grep -oP 'inet \K([0-9]+\.){3}[0-9]+'"
```
Comment (robfallows) [inline deploy/v2/ansible/roles/test-failover/vars/main.yml:37]:
```suggestion
    db_ip: "ip addr show eth1 | grep -m1 inet | grep -oP 'inet \\K([0-9]+\\.){3}[0-9]+'"
```

File: deploy/v2/ansible/roles/test-failover/tasks/1_check_cluster_available.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+- name: Check the secondary (DEMOTED) node has status SOK
+  shell: "{{ cluster_availability_cmds[ansible_os_family].secondary_sok }}"
+  changed_when: false
+  failed_when: false
+  register: check_secondary_status_sok
+
+- name: Set fact for cluster availability
+  set_fact:
+    failover_cluster_available: "{{ check_cluster_available.rc == 0) and check_secondary_status_sok.rc == 0) }}"
```
Comment (robfallows) [inline deploy/v2/ansible/roles/test-failover/tasks/1_check_cluster_available.yml:17]:
```suggestion
    failover_cluster_available: "{{ (check_cluster_available.rc == 0) and (check_secondary_status_sok.rc == 0) }}"
```

File: deploy/v2/ansible/roles/test-failover/tasks/2_gather_cluster_facts.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+---
+
+- name: Dynamically get the Master node (PROMOTED) hostname
```
Comment (robfallows) [inline deploy/v2/ansible/roles/test-failover/tasks/2_gather_cluster_facts.yml:3]: Is it better to use a name like `Ensure Master node (PROMOTED) hostname is known`?

File: deploy/v2/ansible/test_failover.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
+
+- hosts: hanadbnodes
+  become: yes
+  become_user: root
+  roles:
+    - role: test-failover
+      vars:
+        failover_type: "{{ failover_test_type | lower }}"
```
Comment (robfallows) [inline deploy/v2/ansible/test_failover.yml:9]: When it's high availability?

Comment (peterchoo) [reply deploy/v2/ansible/test_failover.yml:9]: Perhaps, but the code itself will not attempt to trigger a failover if the clustering software isn't available.

The code was designed to discover everything,s o only the type of test provided is required.

**Pull request merged**